sdk/src/main/scala/com/horizen/certificatesubmitter/CertificateSubmitter.scala

```diff
@@ -41,37 +50,52 @@ class CertificateSubmitter
 
   lazy val mainchainApi = new RpcMainchainApi(settings)
 
-  protected def viewAsync(): Future[SidechainNodeView] = {
-    def f(v: SidechainNodeView) = v
-
-    (sidechainNodeViewHolderRef ? GetDataFromCurrentSidechainNodeView(f))
-      .mapTo[SidechainNodeView]
-  }
+  protected def submitCertificate(sidechainNodeView: View): SubmitResult = {
+    try {
 
-  protected def trySubmitCertificate: Unit = {
-    viewAsync().onComplete {
-      case Success(sidechainNodeView) => {
-        val withdrawalEpochInfo = sidechainNodeView.getNodeState.getWithdrawalEpochInfo
-        if (withdrawalEpochInfo.epoch > 0 &&
-            WithdrawalEpochUtils.canSubmitCertificate(withdrawalEpochInfo, params)) {
-          val withdrawalRequests = sidechainNodeView.getNodeState.getWithdrawalRequests(withdrawalEpochInfo.epoch - 1)
-          if (withdrawalRequests.isPresent) {
-            val mcEndEpochBlockHashOpt = sidechainNodeView.getNodeHistory
-              .getMainchainBlockReferenceInfoByMainchainBlockHeight(
-                params.mainchainCreationBlockHeight +
-                  withdrawalEpochInfo.epoch * params.withdrawalEpochLength - 1)
-            mainchainApi.sendCertificate(
-              CertificateRequestCreator.create(
-                withdrawalEpochInfo.epoch - 1,
-                mcEndEpochBlockHashOpt.get().getMainchainBlockReferenceHash,
-                withdrawalRequests.get().asScala,
-                params
-              )
+      val withdrawalEpochInfo = sidechainNodeView.state.getWithdrawalEpochInfo
+      if (withdrawalEpochInfo.epoch > 0 &&
+        WithdrawalEpochUtils.canSubmitCertificate(withdrawalEpochInfo, params)) {
+        val withdrawalRequests = sidechainNodeView.state.getWithdrawalRequests(withdrawalEpochInfo.epoch - 1)
+        if (withdrawalRequests.isPresent) {
```

Comment:
Copy from previous conversation:
Ok, I get it.
Main logical problem here, that sidechainNodeView.state.getWithdrawalRequests actually means: return me requests for that epoch if the cert is not present yet.
sidechainNodeView.state.getUnprocessedWithdrawalRequests is much clear.
getWithdrawalRequests must return the seq in any case.

As for me, just retrieving the boolean flag state.containsCertificateForEpoch(epoch) will simplify a lot, instead of having getWithdrawalRequests and getUnprocessedWithdrawalRequests or just one with mixed logic.